### ISSUE TITLE
Nonce verification doesn't play nice with input var validation

### ISSUE THREAD

**Issue by westonruter** [WordPress-Coding-Standards/WordPress-Coding-Standards#414]
Take this for example:

```php
$nonce = isset( $_POST['nonce'] ) ? ( sanitize_text_field( wp_unslash( $_POST['nonce'] ) ) ) : ''; // wpcs: input var okay

if ( ! wp_verify_nonce( $nonce, 'foo-action' ) ) {
        wp_send_json_error( __( 'Not authorized' ) );
}
```

PHPCS flags the first statement for not verifying with a nonce, but if nonce verification _is_ done first, then it flags the nonce verification for not using a validated input var.

The simplest solution is to just change the approach:

```php
if ( ! check_ajax_referer( 'foo-action', 'nonce', false ) ) {
	wp_send_json_error( __( 'Not authorized' ) );
}
```

But it seems there is no way to use `wp_verify_nonce()` given the current state of the nonce verification check.

**Comment by JDGrimes** [comment]
This will work:

```php
	if ( ! isset( $_POST['nonce'] ) || ! wp_verify_nonce( sanitize_key( $_POST['nonce'] ), 'foo-nonce' ) ) {
		wp_send_json_error( __( 'Not authorized' ) );
	}
```

I think that there are two issues here.

First, `wp_unslash()` is being flagged, when it should be treated like a sanitizing function here. However, you should be able to use `sanitize_key()` to sanitize your nonce, because nonces don't contain special characters (although they are pluggable). The sniffs are smart enough to know that `sanitize_key()` doesn't need to be accompanied by `wp_unslash()`, so you could replace `sanitize_text_field( wp_unslash( $_POST['nonce'] ) )` with `sanitize_key( $_POST['nonce'] )`.

Second, the sniff is being derailed because you are using the ternary operator. It just doesn't support that right now. You would have to use `if` instead:

```php
	if ( ! isset( $_POST['nonce'] ) ) {
		wp_send_json_error( __( 'Not authorized' ) );
	}

	$nonce = sanitize_key( $_POST['nonce'] );
	if ( ! wp_verify_nonce( $nonce, 'foo-nonce' ) ) {
		wp_send_json_error( __( 'Not authorized' ) );
	}
```

I'm going to leave support for the ternary operator for a separate ticket. Let's focus on the bug with `wp_unslash()`.

**Comment by JDGrimes** [comment]
I'm leaning toward `wontfix` on this. The `sanitize_key()` function seems like the right one to use to sanitize an nonce, and in that case `wp_unslash()` isn't needed.

**Comment by westonruter** [comment]
k, let's close
